### ISSUE TITLE
BUG: Trouble fixing shebangs

### ISSUE THREAD

**Issue by jakirkham** [conda/conda-build#889]
Running into some sort of [ASCII/unicode bug]( https://travis-ci.org/conda-forge/vigra-feedstock/jobs/124919857#L953-L977 ) in `conda-build` with on Python 3.5 with `conda-build` version [1.20.1]( https://travis-ci.org/conda-forge/vigra-feedstock/jobs/124919857#L218 ) on Mac.

```
updating shebang: bin/vigra-config
An unexpected error has occurred, please consider sending the
following traceback to the conda GitHub issue tracker at:

    https://github.com/conda/conda-build/issues

Include the output of the command 'conda info' in your report.
Traceback (most recent call last):
  File "/Users/travis/miniconda3/bin/conda-build", line 5, in <module>
    sys.exit(main())
  File "/Users/travis/miniconda3/lib/python3.5/site-packages/conda_build/main_build.py", line 208, in main
    args_func(args, p)
  File "/Users/travis/miniconda3/lib/python3.5/site-packages/conda_build/main_build.py", line 493, in args_func
    args.func(args, p)
  File "/Users/travis/miniconda3/lib/python3.5/site-packages/conda_build/main_build.py", line 438, in execute
    include_recipe=args.include_recipe)
  File "/Users/travis/miniconda3/lib/python3.5/site-packages/conda_build/build.py", line 515, in build
    post_build(m, sorted(files2 - files1))
  File "/Users/travis/miniconda3/lib/python3.5/site-packages/conda_build/post.py", line 332, in post_build
    fix_shebang(f, osx_is_app=osx_is_app)
  File "/Users/travis/miniconda3/lib/python3.5/site-packages/conda_build/post.py", line 76, in fix_shebang
    fo.write(new_data)
TypeError: write() argument must be str, not bytes
```

**Comment by stuarteberg** [comment]
Just ran into this on Linux, too.

```
$ conda --version
conda 4.0.5
$ conda build --version
conda-build 1.20.1
```

**Comment by stuarteberg** [comment]
Downgrading to conda-build 1.20.0 fixes the problem.  Presumably, d398611fde9395b8f921d4ada6ae3e245f82c71f is to blame.  Do you have any thoughts, @brentp?

**Comment by brentp** [comment]
yes, that's me. looks like we can just add .decode('ascii')  to the write. trying to figure out how to trigger this error in the existing tests...

**Comment by stuarteberg** [comment]
I don't know if this will be useful, but the particular file it failed on is the following.  Note that this is what the file looks like *after* it is patched by `fix_shebang()`.  (Also, this version is from my Mac, not Linux.)

```python
#!/miniconda/envs/ilastik-devel/bin/python

from optparse import OptionParser
import sys

hasFFTW = bool('/miniconda/envs/ilastik-devel/lib/libfftw3.dylib')

parser = OptionParser()

parser.add_option("--version", action = 'store_true',
                  help = "output version (1.10.0)")
#parser.add_option("--target", action = 'store_true',
#                 help = "output platform this VIGRA library was configured for")
parser.add_option("--impex-lib", "--libs", action = 'store_true',
                  help = "output flags for linking libvigraimpex")
parser.add_option("--fftw-lib", action = 'store_true',
                  help = "output flags for linking libfftw"
                  + " (unavailable)" if not hasFFTW else "")
#parser.add_option("--rfftw-lib", action = 'store_true',
#                 help = "output flags for linking librfftw and libfftw")
parser.add_option("--cppflags", action = 'store_true',
                  help = "output include flags for vigra"
                  + " and fftw" if hasFFTW else "")
parser.add_option("--include-path", "--includepath", action = 'store_true',
                  help = "output path to VIGRA includes")
parser.add_option("--docdir", action = 'store_true',
                  help = "output path to VIGRA documentation")

(op, args) = parser.parse_args()

if len(sys.argv) < 2:
    sys.stderr.write("ERROR: no parameters given.\n")
    parser.print_help()
    sys.exit(1)

# --------------------------------------------------------------------

import os.path

# don't output -L flags for the following dirs (assumed to be compiler
# built-in):
standardLibDirs = ["/lib", "/usr/lib", "/lib64", "/usr/lib64"]

libExt = ".so" if not sys.platform == "darwin" else ".dylib"

def filename2ldflags(libFilename):
    dir, fn = os.path.split(libFilename)
    if not (fn.startswith("lib") and fn.endswith(libExt)):
        if libFilename:
            sys.stderr.write("ERROR: don't know how to handle %r!\n" % libFilename)
        return []
    lib = fn[3:-len(libExt)]

    result = []
    if dir not in standardLibDirs:
        result.append("-L" + dir)
    result.append("-l" + lib)
    return result

if op.version:
    print "1.10.0"

if op.cppflags: # was: --cppflags|--cxxincludes|--cxxflags|--cincludes|--cflags
    print '-I/miniconda/envs/ilastik-devel/include'

if op.impex_lib: # was: --impex_lib|--impex-lib|--libs
    ldflags = []

    libDir = '/miniconda/envs/ilastik-devel/lib'
    if libDir not in standardLibDirs:
        ldflags.append('-L' + libDir)

    ldflags.append('-lvigraimpex')

    for flag in 'general;/miniconda/envs/ilastik-devel/lib/libjpeg.dylib;general;/miniconda/envs/ilastik-devel/lib/libpng.dylib;general;/miniconda/envs/ilastik-devel/lib/libz.dylib;general;/miniconda/envs/ilastik-devel/lib/libtiff.dylib;general;/miniconda/envs/ilastik-devel/lib/libhdf5.dylib;general;/miniconda/envs/ilastik-devel/lib/libhdf5_hl.dylib;general;/miniconda/envs/ilastik-devel/lib/libz.dylib;'.split(';'):
        if flag == 'general':
            continue
        for fl in filename2ldflags(flag):
            if fl not in ldflags:
                ldflags.append(fl)

    print " ".join(ldflags)

if op.fftw_lib:
    if not hasFFTW:
        sys.stderr.write("VIGRA was configured without FFTW switches, libpath unknown!\n")
        sys.exit(1)
    print " ".join(filename2ldflags('/miniconda/envs/ilastik-devel/lib/libfftw3.dylib'))

if op.include_path: # was: --include_path|--include-path|--includepath
    print '/miniconda/envs/ilastik-devel/include'

if op.docdir:
    print '/miniconda/conda-bld/work/doc'
```

**Comment by brentp** [comment]
I thinks it's with detect_binary_files_with_prefix: true

**Comment by stuarteberg** [comment]
Could be.  This is the recipe:
https://github.com/ilastik/ilastik-build-conda/blob/master/vigra/meta.yaml

**Comment by brentp** [comment]
aye. but reading that is where I got the idea... :|

**Comment by stuarteberg** [comment]
By the way, I'm using the Python 2 version of Miniconda. (Probably should have mentioned that earlier...)

**Comment by stuarteberg** [comment]
Brent fixed this in #892.  I think you can close this issue, @jakirkham.

**Comment by jakirkham** [comment]
I believe you are right, but will need to re-test to verify that this is actually fixed. Will try to find some time to do this. Unfortunately, this is only one bug that crawled into conda-build 1.20.1.

**Comment by ngoldbaum** [comment]
We're hitting this in our CI infrastructure: https://tests.yt-project.org/job/conda_ytdev_linux/91/console

It would be nice if you could cut a new conda release with the fix.

**Comment by jakirkham** [comment]
@ngoldbaum, the problem isn't present in 1.20.0 so you could downgrade for the time being.

**Comment by jakirkham** [comment]
To the best of my knowledge this is fixed in 1.20.2. Closing...
